### PR TITLE
[SPARK-35794][SQL] Allow custom plugin for AQE cost evaluator

### PR DESCRIPTION
**Pull request by c21**
### What changes were proposed in this pull request?

Current AQE has cost evaluator to decide whether to use new plan after replanning. The current used evaluator is `SimpleCostEvaluator` to make decision based on number of shuffle in the query plan. This is not perfect cost evaluator, and different production environments might want to use different custom evaluators. E.g., sometimes we might want to still do skew join even though it might introduce extra shuffle (trade off resource for better latency), sometimes we might want to take sort into consideration for cost as well. Take our own setting as an example, we are using a custom remote shuffle service (Cosco), and the cost model is more complicated. So We want to make the cost evaluator to be pluggable, and developers can implement their own `CostEvaluator` subclass and plug in dynamically based on configuration.

The approach is to introduce a new config to allow define sub-class name of `CostEvaluator` - `spark.sql.adaptive.customCostEvaluatorClass`. And add `CostEvaluator.instantiate` to instantiate the cost evaluator class in `AdaptiveSparkPlanExec.costEvaluator`.

### Why are the changes needed?

Make AQE cost evaluation more flexible.

### Does this PR introduce _any_ user-facing change?

No but an internal config is introduced - `spark.sql.adaptive.customCostEvaluatorClass` to allow custom implementation of `CostEvaluator`.

### How was this patch tested?

Added unit test in `AdaptiveQueryExecSuite.scala`.

### CODE REVIEW

**Comment by c21** [comment]
cc @cloud-fan could you help take a look when you have time? Thanks.

**Comment by cloud-fan** [comment]
does it work well with https://github.com/apache/spark/pull/32816 ?

**Comment by c21** [comment]
> does it work well with #32816 ?

@cloud-fan - I think so. If we decide merge this first, then in https://github.com/apache/spark/pull/32816, we don't need the extra config `spark.sql.adaptive.forceEnableSkewJoin`. Developers/users can set `spark.sql.adaptive.costEvaluatorClass` to `SkewJoinAwareCostEvaluator` and it should work, cc @ulysses-you FYI, thanks.

**Comment by ulysses-you** [comment]
@c21 thank you for ping me.

Not sure it's worth to make cost evaluator as plugin. You mentioned sort (I think it's local sort, isn't it ?), and can you provide a real use case about it ?

**Comment by cloud-fan** [comment]
> Developers/users can set spark.sql.adaptive.costEvaluatorClass to SkewJoinAwareCostEvaluator and it should work

I don't think it's that simple. If force-skew-join-handling is enabled, Spark must use `SkewJoinAwareCostEvaluator`, not a user-specified one.

**Comment by c21** [comment]
> You mentioned sort (I think it's local sort, isn't it ?), and can you provide a real use case about it ?

@ulysses-you - e.g.

```
SortAggregate
- SortMergeJoin
  - Sort(Shuffle(Scan))
  - Sort(Shuffle(Scan))       
```

AQE might change it to

```
SortAggregate
- Sort
  - ShuffledHashJoin
    - Shuffle(Scan)
    - Shuffle(Scan)
```

With our Cosco remote shuffle service, we already implemented the sorted shuffle (`Sort(Shuffle)` where `Sort` and `Shuffle` down in shuffle service side together at same time), and it would be more efficient than doing `Sort` separately in Spark. So a `Sort(Shuffle)` is more efficient than a pair of `Shuffle` and `Sort` in our case. This influences our AQE decision and we have to have a custom cost evaluator. As we can see, we need a separate cost evaluator for forcing skew join and we might have more in the future. Another aspect is Cosco hasn't been open source yet, so we want a clean interface for custom cost evaluator, instead of always maintaining a fork change on our side.

**Comment by c21** [comment]
> I don't think it's that simple. If force-skew-join-handling is enabled, Spark must use SkewJoinAwareCostEvaluator, not a user-specified one.

@cloud-fan - from my checking of https://github.com/apache/spark/pull/32816, it looks like the only logic controlled by the new config `spark.sql.adaptive.forceEnableSkewJoin`, is to choose a different cost evaluator - `SkewJoinAwareCostEvaluator`. My idea is to not introduce the new config, but we can just set `spark.sql.adaptive.costEvaluatorClass` to `SkewJoinAwareCostEvaluator` to enable force skew join.

**Comment by ulysses-you** [comment]
@c21 thanks for the explaination, the example `SortAggregate(SMJ)` to `SortAggregate(SHJ)` seems useful. But about the usage, I agree with @cloud-fan , the boolean config of `forceEnableSkewJoin` is necessary and more easy for user. A class name is a little hack for user if they want to optimize skew join anyway.

**Comment by c21** [comment]
> the boolean config of forceEnableSkewJoin is necessary and more easy for user. A class name is a little hack for user if they want to optimize skew join anyway.

@ulysses-you - sure, I agree with boolean config is more intuitive and easier to use. If we do need the boolean config, we can add special logic in `AdaptiveSparkPlanExec.costEvaluator`, to use `SkewJoinAwareCostEvaluator` when `spark.sql.adaptive.forceEnableSkewJoin` is true, instead of whatever user sets for `spark.sql.adaptive.costEvaluatorClass`. It just a matter of priority between different configs.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:650]: We can make it an optional conf: `spark.sql.adaptive.customCostEvaluatorClass`. If not set, we use the builtin impl.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/costing.scala
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/costing.scala:43]: We can use `Utils.loadExtensions`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/simpleCosting.scala
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/simpleCosting.scala:41]: This can still be an object, if we follow https://github.com/apache/spark/pull/32944/files#r662513062

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (c21) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: @cloud-fan - sure, updated.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/costing.scala
Comment (c21) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/costing.scala]: @cloud-fan - good call, updated.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/simpleCosting.scala
Comment (c21) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/simpleCosting.scala]: @cloud-fan - yeah, updated.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3567]: nit: we don't have to create a method here if it's only called once

File: sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:1904]: nit: `val (_, adaptivePlan) = ...`

File: sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:1903]: does this custom cost evaluator change the query plan? It seems to be the same with the builtin cost evaluator.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (c21) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: @cloud-fan - sure, updated.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala
Comment (c21) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:1903]: @cloud-fan - this evaluator does not change plan, and to be the same with the builtin evaluator for this query. Do we want to come up a different one here? I think this just validates the custom evaluator works.

Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:1903]: SGTM

**Comment by cloud-fan** [comment]
@c21 can you fix the code conflicts?

**Comment by c21** [comment]
@cloud-fan - thanks, just rebased to latest master.

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:134]: Hey, how do we use this? `CostEvaluator` isn't an API.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/costing.scala
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/costing.scala:34]: whole execution package is for internal purpose. I don't think it makes much sense to make it pluggable.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:686]: Why is it an internal configuration?

**Comment by HyukjinKwon** [comment]
@c21, can you at least mark `CostEvaluator` with `@Unstable` API tag? Also please add a note that it is subject to be moved or changed in the near future.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala
Comment (c21) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:134]: @HyukjinKwon - good callout. If the whole execution package is for internal purpose (I think the only public one is around data source, right?), how about moving `Cost` and `CostEvaluator` to `org.apache.spark.sql.catalyst.planning` or `org.apache.spark.sql.catalyst.plans.physical`?

Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:134]: I'm okay with that.

Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:134]: but the cost stuff should live under physical plans .. I would just mark it as `@Unstable` for now and unblock this PR.

Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:134]: hey sorry I left some comments and removed it back. Yes, it makes sense to me.

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:134]: `catalyst` is also an internal package. 

`SparkSessionExtensions` exposes `LogicalPlan` as well, and I think unstable developer API does not have a hard requirement on the package names. I'm fine to add `@Unstable`

Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:134]: Yeah, we can add `@Unstable` for now but I would also add a note that this class is supposed to be moved or changed in the near future.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (c21) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:686]: @HyukjinKwon - I think it can be a non-internal one, updated.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala
Comment (c21) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala:134]: Sounds good, updated with `@Unstable` and added a note comment for class `Cost` and `CostEvaluator`.

**Comment by c21** [comment]
@HyukjinKwon - updated per discussion, and this is ready for review again, thanks.

**Review by HyukjinKwon — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:685]: the only think is that the version has to be 3.3.0 since we cut the branch now. Since this PR won't likely affect anything in the main code, I am okay with merging to 3.2.0 either tho. I will leave it to @cloud-fan

**Review by HyukjinKwon — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:685]: 3.2 is the first version that enables AQE by default, and this seems to be a useful extension. Let's include it in 3.2.

**Comment by cloud-fan** [comment]
thanks, merging to master/3.2!

**Comment by c21** [comment]
Thank you @cloud-fan and @HyukjinKwon for review!

**Pull request closed without merging**
